[blobfish465] iP

blobfish465 opened the pull request:
### Blob Task Manager 

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Blob is:

- Text-based
- User-friendly
- ~~Fast~~ _SUPER_ FAST to use 

Here's how to get started:

1. Clone the repository from [here](https://github.com/blobfish465/ip)
2. Compile the Java files.
3. Run the application and start managing your tasks!
4. Add your task
5. Let it manage your task for you! :smile:

And it is **FREE!**

Features:

- [X] Managing tasks: Add, delete, and mark tasks as done.
- [X] Time-bound tasks: Set deadlines and events with specific times. 
- [X] Search: Quickly find tasks by keywords.
- [ ] Reminders (Coming Soon!)


If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public static void main(String[] args) {
    new Blob("data/Blob.txt").run();
}
```

Enjoy managing your tasks with Blob! :memo:

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

blobfish465 committed:
Rename Duke to Blob, Add greeting and farewell

blobfish465 committed:
Add echo and bye command

blobfish465 committed:
Add store task and list command

blobfish465 committed:
Add ability to mark task as done or unmark task

blobfish465 committed:
Add ToDos, Events, Deadlines classes

blobfish465 committed:
Add automated test UI testing and new line after every message

blobfish465 committed:
Add Error Handling and Exceptions

blobfish465 committed:
Add delete command with testing and import array list from java collections

blobfish465 committed:
Add save functionality

loads data from hard disk when chatbot starts up, tasks are saved and read from blob.txt file

blobfish465 committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Add save functionality

blobfish465 committed:
Add Dates and Times feature to deadline and events

Input format for deadlines: deadline submit report /by 2025-03-01 1700
Input format for events: event team meeting /from 2025-03-01 1700 /to 2025-03-01 2000
Print output format: Mar 1 2025, 5:00 PM

blobfish465 committed:
Revert "Add Dates and Times feature to deadline and events"

This reverts commit 288ad13e418e7288c186bd259f9de3f81be9e13b.

blobfish465 committed:
Reapply "Add Dates and Times feature to deadline and events"

This reverts commit 8598e3a7aa270b08dfbb57a2d268886c1b7e9731.

blobfish465 committed:
Reapply "Add Dates and Times feature to deadline and events"

This reverts commit 8598e3a7aa270b08dfbb57a2d268886c1b7e9731.

blobfish465 committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Reapply "Add Dates and Times feature to deadline and events"

blobfish465 committed:
Refactor code to use more OOP

Ui, Storage, Parser, Tasklist classes

Command interface and command classes(Delete Command ExitCommand, ListCommand, MarkCommand, UnmarkCommand)

blobfish465 committed:
Organize the classes into suitable java packages

blobfish465 committed:
Change name of txt file

blobfish465 committed:
Merge commit '81a9c537afe8492675eb253968939364d65db091'

* commit '81a9c537afe8492675eb253968939364d65db091':
  build.gradle: Prevent generating a second JAR file
  Bump gradle and lib version
  Add Gradle support

blobfish465 committed:
Set up gradle

build, run and test with gradle successful

blobfish465 committed:
Add JUnit test for parser and TaskList classes

Parsing valid todo input should return Add Command while invalid input returns UnknownCommandException

Adding task into TaskList should be done correctly with valid input

Passed

blobfish465 committed:
Add JavaDoc comments to all non-private classes/methods

blobfish465 committed:
Merge branch 'A-JavaDoc'

* A-JavaDoc:
  Add JavaDoc comments to all non-private classes/methods

blobfish465 committed:
Tweak the code to comply with coding standard

blobfish465 committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Tweak the code to comply with coding standard

# Conflicts:
#	src/main/java/blob/Blob.java
#	src/main/java/blob/TaskList.java
#	src/main/java/blob/command/ExitCommand.java
#	src/main/java/blob/model/Event.java
#	src/main/java/blob/model/Task.java
#	src/main/java/blob/parser/Parser.java
#	src/main/java/blob/ui/Ui.java

blobfish465 committed:
Resolve some merge conflicts

blobfish465 committed:
Add find functionality

Give users a way to find a task by searching for a keyword in the task description.

blobfish465 committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add find functionality

# Conflicts:
#	src/main/java/blob/TaskList.java
#	src/main/java/blob/model/Task.java
#	src/main/java/blob/parser/Parser.java
#	src/main/java/blob/ui/Ui.java

blobfish465 committed:
Merge commit '5fcb0d566b77b7bdd30484342277fc66422a6836' into branch-Level-7

* commit '5fcb0d566b77b7bdd30484342277fc66422a6836':

blobfish465 committed:
Merge branch 'refs/heads/branch-Level-7'

# Conflicts:
#	data/blob.txt
#	src/main/java/Blob.java
#	src/main/java/Deadline.java
#	src/main/java/Event.java
#	src/main/java/Task.java
#	text-ui-test/EXPECTED.TXT

blobfish465 committed:
Add some command exceptions

blobfish465 committed:
Add javadoc comments to ListCommand.java

Add throws NoTaskException if the list does not contain any task

blobfish465 committed:
Update build.gradle to include javaFX dependencies

Add some implementation groups to set up java fx

blobfish465 committed:
Add GUI to the chatbot using JavaFX

Add a DialogBox and a MainWindow controller along with its fxml files in the view folder.

Edit Ui to use appendMessage method instead of printing on console, to return the response string to blob and then MainWindow.

blobfish465 committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Add GUI to the chatbot using JavaFX
  Update build.gradle to include javaFX dependencies

blobfish465 committed:
Add javadoc comments to the files involve in level 10

Add javadoc comments for Blob.java, Launcher.java, Main.java, Command.java, DialogBox.java, MainWindow.java and Ui.java.

blobfish465 committed:
Create new JAR file bundling JavaFX third party library into the jar file

Add archiveFileName to build.gradle and enable assertions in build.gradle for future increment

yuqiannemo commented on src/main/java/blob/Main.java:
```diff
@@ -0,0 +1,51 @@
+import javafx.scene.Scene;
+import javafx.scene.layout.AnchorPane;
+import javafx.stage.Stage;
+
```
The ordering of import statements LGTM.

yuqiannemo commented on src/main/java/blob/command/Command.java:
```diff
@@ -0,0 +1,40 @@
+ * through {@link Ui}, and may interact with {@link Storage} to persist changes.
+ */
+public interface Command {
+
```
Should this line break (line 16) be removed?

yuqiannemo commented on src/main/java/blob/model/Task.java:
```diff
@@ -0,0 +1,116 @@
+        String description = parts[2].trim();
+        Task task = null;
+
+        switch (type) {
```
The coding standard for "switch" statements are closely followed!

yuqiannemo commented on src/main/java/blob/command/AddCommand.java:
```diff
@@ -0,0 +1,49 @@
+     * @param tasks The task list to which the task will be added.
+     * @param ui The UI to interact with the user.
+     * @param storage The storage used to save the updated task list.
+     */
```
Nice JavaDoc added! Also, when writing Javadocs for overridden methods, the **@inheritDoc tag** can be used to reuse the header comment from the parent method but with further modifications e.g., when the method has a slightly different behavior from the parent method (this applies to other subclasses of Command class too).

yuqiannemo commented on src/main/java/blob/TaskList.java:
```diff
@@ -0,0 +1,95 @@
+package blob;
```
The naming LGTM.

blobfish465 committed:
Add assertions to ensure code integrity

Introduce assertions across various classes

Several parts of the project lacked assertions for checking the internal state
and assumptions. Adding these assertions helps catch logical errors during
development and ensures that the program's assumptions hold true under various
conditions.

Implemented changes include:
* `TaskList.java`: Assertions to ensure no null tasks are added and index
  validations for task operations.
* `Event.java`, `Task.java`, `ToDo.java`: Assertions to verify correct data
  handling and state consistency.

These additions are critical for maintaining robustness in the system's core
functionalities and prevent runtime errors caused by incorrect internal states.

By enforcing these constraints, we aim to make the debugging process more
efficient and reduce potential faults in production.

blobfish465 committed:
Merge pull request #1 from blobfish465/branch-A-Assertions

Add assertions to ensure code integrity

blobfish465 committed:
Merge branch 'master' of https://github.com/blobfish465/ip

* 'master' of https://github.com/blobfish465/ip:
  Add assertions to ensure code integrity

blobfish465 committed:
Introduce constants and modularize GUI initialization

- In Event.java, centralized the date parsing mechanism to enhance code
  readability and maintainability. Defined constants for date patterns to
  replace magic strings, thus making the code more understandable and less
  error-prone.

- Refactored handleUserInput in MainWindow to split the logic into two
  distinct methods: createUserDialog and createBlobDialog. This separation
  clarifies the process flow and makes the method responsibilities clearer.

These changes are aimed at improving the maintainability and readability of
the codebase, ensuring that future modifications can be made more easily
and with less risk of introducing errors.

blobfish465 committed:
Merge pull request #2 from blobfish465/branch-A-CodeQuality

Introduce constants and modularize GUI initialization

blobfish465 committed:
Merge branch 'master' of https://github.com/blobfish465/ip

* 'master' of https://github.com/blobfish465/ip:
  Introduce constants and modularize GUI initialization

blobfish465 committed:
Set up CI with GitHub Actions and adjust assertion handling

Set up Continuous Integration using GitHub Actions to automate the
build and test processes across multiple platforms. The CI pipeline
is configured to run on push and pull request events for Ubuntu,
macOS, and Windows environments. Additionally, streamline error
handling by replacing certain assertions with exceptions to better
capture user errors during runtime.

Changes:
- Added a new `.github/workflows/gradle.yml` file to define the CI
  process, ensuring builds and tests are automatically performed.
- Removed assertions checking the task index bounds in `deleteTask`
  and `getTask` methods of the TaskList class. These assertions were
  inappropriate for handling user errors and have been replaced by
  exception handling to throw an `IndexOutOfBoundsException` if the
  provided index is out of the valid range.
- Moved the assertion for `startDateTime` and `endDateTime` in the
  Event class to the constructor. This ensures that the date-time
  objects are consistently set or null right after object
  instantiation, enhancing data integrity and error detection during
  the object creation phase.

This commit improves the robustness of the system by ensuring that
user errors are handled through exceptions, aligning with best
practices for error handling. The CI setup introduces automation
that aids in maintaining code quality across multiple platforms,
streamlining development and testing workflows.

blobfish465 committed:
Modify order of import statements to adhere to java coding standards

Modified the order of import statements for various files to adhere to java coding standards

Ensure the ordering of import statements is consistent to make it easier to browse the list of imports and determine the dependencies

blobfish465 committed:
Edit set up repository in gradle.yml file

uses: actions/checkout@v2 instead of uses: actions/checkout@master

blobfish465 committed:
Add support for DoWithinPeriodTask extension

Introduce the 'Period Task' to provide support for managing tasks that need to be done within a certain period

'Period Task' allows users to define tasks that have a start and end date. This feature is essential for tasks that are not bound to a single deadline but must be completed within a certain timeframe.

Input to CLI example: period Collect certificate /between 2025-01-15 /and 2025-01-25
Output to storage: P | 0 | Collect certificate | 2025-01-15 | 2025-01-25

Changes include:
* Creation of `PeriodTask` class extending the `Task` base class.
* Addition of parsing logic in the `Parser` class to recognize 'period' commands and construct `PeriodTask` instances.
* Update `toString` and `toFileFormat` methods in `PeriodTask` for appropriate display and storage.

The implementation follows the established code quality guidelines by maintaining readability, minimizing code duplication, and ensuring clear separation of concerns within the `Parser` logic.

This feature enhances the application's utility by allowing more flexible task management.

weiwong834 reviewed (commented):
Code looks good! No naming issues found

blobfish465 committed:
Set application title to 'Blob' in GUI window

Add the product name 'Blob' to the title bar of the main application window.

This change enhances user recognition and strengthens branding by displaying the application name prominently in the GUI window's title bar. It provides users with immediate context about the application they are interacting with.

By setting the stage title in the start method, we ensure that every instance of the application window reflects the product name, thereby maintaining consistency throughout the user experience.

This modification aligns with our efforts to improve the GUI's appearance and usability, making the application feel more professional and integrated.

blobfish465 committed:
Add user guide and GUI screenshot to docs folder

Add essential documentation for the Blob app

- Include Ui.png, a screenshot of the full GUI window, in the docs folder. This image showcases the application's interface, emphasizing its ease of use and user-friendly design.

- Add a User Guide (UserGuide.md) to the docs folder. The guide provides new users with instructions on how to effectively use the Blob app, covering basic functionalities like task creation, modification, and interaction with the chatbot.

These additions will help new users understand and navigate the app more effectively, enhancing their overall experience.
